convex/util/ollama.ts

```diff
@@ -0,0 +1,465 @@
+      };
+    }
+    return (await result.json()) as CreateEmbeddingResponse;
+  });
+  if (json.data.length !== texts.length) {
+    console.error(json);
+    throw new Error('Unexpected number of embeddings');
+  }
+  const allembeddings = json.data;
+  allembeddings.sort((a, b) => a.index - b.index);
+  return {
+    embeddings: allembeddings.map(({ embedding }) => embedding),
+    usage: json.usage.total_tokens,
+    retries,
+    ms,
+  };
+}
+
+export async function fetchEmbedding(text: string) {
+  const { embeddings, ...stats } = await fetchEmbeddingBatch([text]);
+  return { embedding: embeddings[0], ...stats };
+}
+
+export async function fetchModeration(content: string) {
+  checkForAPIKey();
+  const { result: flagged } = await retryWithBackoff(async () => {
+    const result = await fetch('https://api.openai.com/v1/moderations', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+        Authorization: 'Bearer ' + process.env.OPENAI_API_KEY,
+      },
+
+      body: JSON.stringify({
+        input: content,
+      }),
+    });
+    if (!result.ok) {
+      throw {
+        retry: result.status === 429 || result.status >= 500,
+        error: new Error(`Embedding failed with code ${result.status}: ${await result.text()}`),
+      };
+    }
+    return (await result.json()) as { results: { flagged: boolean }[] };
+  });
+  return flagged;
+}
+
+const checkForAPIKey = () => {
+  if (!process.env.OPENAI_API_KEY) {
+    throw new Error(
+      'Missing OPENAI_API_KEY in environment variables.\n' +
+        'Set it in the project settings in the Convex dashboard:\n' +
+        '    npx convex dashboard\n or https://dashboard.convex.dev',
+    );
+  }
+};
+
+// Retry after this much time, based on the retry number.
+const RETRY_BACKOFF = [1000, 10_000, 20_000]; // In ms
+const RETRY_JITTER = 100; // In ms
+type RetryError = { retry: boolean; error: any };
+
+export async function retryWithBackoff<T>(
+  fn: () => Promise<T>,
+): Promise<{ retries: number; result: T; ms: number }> {
+  let i = 0;
+  for (; i <= RETRY_BACKOFF.length; i++) {
+    try {
+      const start = Date.now();
+      const result = await fn();
+      const ms = Date.now() - start;
+      return { result, retries: i, ms };
+    } catch (e) {
+      const retryError = e as RetryError;
+      if (i < RETRY_BACKOFF.length) {
+        if (retryError.retry) {
+          console.log(
+            `Attempt ${i + 1} failed, waiting ${RETRY_BACKOFF[i]}ms to retry...`,
+            Date.now(),
+          );
+          await new Promise((resolve) =>
+            setTimeout(resolve, RETRY_BACKOFF[i] + RETRY_JITTER * Math.random()),
+          );
+          continue;
+        }
+      }
+      if (retryError.error) throw retryError.error;
+      else throw e;
+    }
+  }
+  throw new Error('Unreachable');
+}
+
+// Lifted from openai's package
+export interface LLMMessage {
```

Comment:
Curious, is this used, or is the prompt built up from the messages in the `conversation.ts` file?

Happy to help with the prompt formatting/building stuff since I know Ollama's API doesn't have an easy way to send messages (yet – this is something we're looking at making easier :-)